### PR TITLE
fix(moe): correct unified fuzzer references

### PR DESCRIPTION
**Pull request by feih-nv**
## 📌 Description

Fix two false failures exposed by the unified MoE fuzzer on GB200:

- Increase the MxInt4 absolute tolerance fraction from `0.06` to `0.065`, calibrated using the full 160-seed GB200 sweep.
- Correct DeepSeekV3 reference routing to mask experts outside selected groups with `-inf` instead of zero. Zero masking could incorrectly select an out-of-group expert when a valid in-group biased score was negative.
- Apply the routing correction to both unified and legacy references.
- Add focused regression coverage for negative biased scores.

No production kernel behavior is changed.

## 🔍 Related Issues

- Follow-up to #4475

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All relevant tests are passing.

Validation performed:

- Focused DeepSeekV3 negative-score regression: passed.
- Seed 18 with full checks and `atol_frac=0.065`: passed.
- GB200 MxInt4 160-seed sweep: `17 passed, 176 skipped`.
- Seed 153 with corrected reference masking: passed.

## Reviewer Notes

Please focus on:

- Whether `0.065` provides an appropriate margin for MxInt4 accuracy.
- The use of `-inf` masking to enforce DeepSeekV3 selected-group membership.
- Keeping the unified and legacy routing references semantically consistent.

## Summary by CodeRabbit

* **Bug Fixes**
  * Corrected mixture-of-experts routing so unselected groups are reliably excluded, including when routing scores are negative.
  * Improved routing behavior for DeepSeek-V3 and related fused implementations.
  * Adjusted MxInt4 validation tolerance based on expanded calibration coverage.

* **Tests**
  * Added regression coverage for negative-score routing and expert-group selection.
  * Added CPU validation confirming selection of only the highest-scoring expert group.

### CODE REVIEW

**Comment by feih-nv** [comment]
@flashinfer-bot run

File: tests/model_optimizations/test_dsv3_fused_routing.py [hunk 1]
```diff
@@ -202,12 +202,14 @@ def __init__(
 
         for token_idx in range(self.num_tokens):
             # Create mask for selected groups
-            group_mask = torch.zeros(n_group, dtype=torch.float32, device=self.device)
-            group_mask[self.ref_group_indices[token_idx]] = 1.0
+            group_mask = torch.zeros(n_group, dtype=torch.bool, device=self.device)
```
Comment (aleozlx) [inline tests/model_optimizations/test_dsv3_fused_routing.py:205]: i'm curious what difference did this make?

**Review by aleozlx — APPROVED** [review]

**Comment by aleozlx** [comment]
## PR Review Screening

**CI verdict:** ✅ auto-run ok
**Review category:** `bot` (rule fired: no live/human rule matched — tests-only reference correction with regression coverage)
**Blocking checks:** none
**Release blocker:** no — tests-only change; the corrected "wrong results" are in test references, not shipped library code (C3.4 library-side requirement)
**Early stop:** no

### Security
| Q | Answer | Evidence |
|---|--------|----------|
| S1 injection/supply-chain | no | — |
| S2 template overwritten | no | — |
| S3 template obligations | met | All headers present, 5 boxes checked, validation results listed, tests added in diff |
| S4 agent-directing text | no | — |

### Packaging
| Q | Answer | Evidence |
|---|--------|----------|
| C1.1 dependency bump | no | — |
| C1.2 public API changes | no | 4 files, all under `tests/`; no `flashinfer/` changes |
| C1.3 AOT/trace registration | n/a | no new modules or APIs |

### Presentation
| Q | Answer | Evidence |
|---|--------|----------|
| C2.1 perf claim backed | n/a | no perf claim; tolerance change is calibrated (160-seed GB200 sweep cited in body and code comment) |

### Implementation
| Q | Answer | Evidence |
|---|--------|----------|
| C3.1 experimental-track | no | — |
| C3.2 shared internals touched | no | test-side reference routing only (`tests/moe/trtllm_gen_fused_moe_utils.py:2315`, `tests/moe/test_unified_moe_fuzz.py:1606`) |
| C3.3 tests match change | yes | 3 focused negative-score regression tests added, one per corrected reference |
| C3.4 release blocker | no | fixes false fuzzer failures (test references), no library-code change |

**Notes for the maintainer**
- The same `-inf`-masking correction is applied consistently across unified, legacy, and DSv3 references — the reviewer-notes ask (semantic consistency) is satisfied structurally.
- `atol_frac` 0.06→0.065 is SM100-calibrated only; the code comment flags SM103/SM107 as uncalibrated.

<sub>Generated by flashinfer-pr-screen · rubric: docs/code_review_guidance.md · not a code review · AI screening can make mistakes — a maintainer's judgment supersedes this report.</sub>

File: tests/model_optimizations/test_dsv3_fused_routing.py (hunk 1 above)
Comment (feih-nv) [reply tests/model_optimizations/test_dsv3_fused_routing.py:205]: The type changed from `float32 (1.0/0.0)` to `bool (True/False)` because this tensor is a mask that represents group membership, not a numerical scaling factor. The type change itself does not alter which groups are marked as selected. So I would say no difference here.

The important difference is how the mask is applied. For example, previously:
```
Selected score: -0.2 × 1.0 = -0.2
Excluded score:  0.8 × 0.0 =  0.0
```
topk could incorrectly select the excluded expert because 0.0 > -0.2.

A boolean mask allows us to use:
```
scores.masked_fill(~expert_mask, float("-inf"))
```
producing:
```
Selected score: -0.2
Excluded score: -inf
```
Selected experts retain their original scores, while excluded experts become ineligible. This fixes a latent bug in the test/reference implementation exposed by seed 153; the production kernel was correct.
